util/src/Pos/Util/Log/Scribes.hs

```diff
@@ -1,75 +1,110 @@
+                     initializeRotator)
+
 import           System.IO (BufferMode (LineBuffering), Handle,
-                     IOMode (WriteMode), hFlush, hSetBuffering, stderr, stdout)
-import           System.IO.Unsafe (unsafePerformIO)
-
--- | global lock for file Scribes
-{-# NOINLINE lock #-}
-lock :: MVar ()
-lock = unsafePerformIO $ newMVar ()
-
-prtoutException :: Exception e => FilePath -> e -> IO ()
-prtoutException fp e = do
-    putStrLn $ "error while opening log @ " ++ fp
-    putStrLn $ "exception: " ++ displayException e
-
--- | create a katip scribe for logging to a file (JSON)
-mkJsonFileScribe :: FilePath -> FilePath -> Severity -> Verbosity -> IO Scribe
-mkJsonFileScribe bp fp s v = do
-    h <- catchIO (openFile (bp </> fp) WriteMode) $
-            \e -> do
-                prtoutException (bp </> fp) e
-                return stdout    -- fallback to standard output in case of exception
-    hSetBuffering h LineBuffering
-    locklocal <- newMVar ()
+                     IOMode (WriteMode), hClose, hSetBuffering, stderr, stdout)
+
+-- | create a katip scribe for logging to a file in JSON representation
+mkJsonFileScribe :: RotationParameters -> Internal.FileDescription -> Severity -> Verbosity -> IO Scribe
+mkJsonFileScribe rot fdesc s v = do
+    mkFileScribe rot fdesc formatter False s v
+  where
+    formatter :: LogItem a => Handle -> Bool -> Verbosity -> Item a -> IO Int
+    formatter hdl _ v' item = do
+        let tmsg = encodeToLazyText $ itemJson v' item
+        TIO.hPutStrLn hdl tmsg
+        return $ length tmsg
+
+-- | create a katip scribe for logging to a file in textual representation
+mkTextFileScribe :: RotationParameters -> Internal.FileDescription -> Bool -> Severity -> Verbosity -> IO Scribe
+mkTextFileScribe rot fdesc colorize s v = do
+    mkFileScribe rot fdesc formatter colorize s v
+  where
+    formatter :: LogItem a => Handle -> Bool -> Verbosity -> Item a -> IO Int
+    formatter hdl colorize' v' item = do
+        let tmsg = toLazyText $ formatItem colorize' v' item
+        TIO.hPutStrLn hdl tmsg
+        return $ length tmsg
+
+-- | create a katip scribe for logging to a file
+--   and handle file rotation within the katip-invoked logging function
+mkFileScribe
+    :: RotationParameters
+    -> Internal.FileDescription
+    -> (forall a . LogItem a => Handle -> Bool -> Verbosity -> Item a -> IO Int)      -- format and output function, returns written bytes
+    -> Bool                              -- whether the output is colourized
+    -> Severity -> Verbosity
+    -> IO Scribe
+mkFileScribe rot fdesc formatter colorize s v = do
+    trp <- initializeRotator rot fdesc
+    scribestate <- newMVar trp    -- triple of (handle), (bytes remaining), (rotate time)
+    -- sporadically remove old log files - every 10 seconds
+    cleanup <- mkAutoUpdate defaultUpdateSettings { updateAction = cleanupRotator rot fdesc, updateFreq = 10000000 }
```

Comment:
cleanup of old files which outnumber the limit set in the configuration file are deleted from the filestystem, at most every 10 seconds. (updateFreq is actually the sleep time).